### ISSUE TITLE
Mp-JWT Token compliance issues.

### ISSUE THREAD

**Issue by BarDweller** [OpenLiberty/open-liberty#6038]
Trying to code a Spring app to process MP-JWT tokens, discovered the tokens issued/accepted by Liberty are not  MP-JWT spec compliant. 

Documentation for the required claims is at
https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/interoperability.asciidoc#minimum-mp-jwt-required-claims

Where it currently states that kid & jti are among the required claims. 

If you use JwtSso to protect an app, the token injected via @Inject JsonWebToken is missing `kid` from it's header, and `jti` from it's claims, and is thus non-compliant to mp-jwt. (although you can make Liberty add if you know to add `jti="true"` as part of the jwtBuilder configuration, its not a defualt)

The *non-compliant* token is also *accepted* by liberty as authorization to access apps configured to be protected by the mp-jwt feature. 

(Discovered while building a Spring authentication filter to accept mp-jwt tokens, Spring filter being written to MP-JWT spec rejected the Liberty generated tokens as invalid.)

In general the MP-JWT spec seems very weak when compared to regular RFC style specs. In discussion of these issues, it has been suggested that jti, kid, and even groups, should all have been optional, or that the spec should be read as just "recommendations". The spec needs to be clear in what constitutes a valid mp-jwt token, to the same level of detail as the jwt spec it builds upon. (Eg, does a `groups` claim of an empty list mean the same as the absence of a `groups` claim)

In my view, `kid` should be allowed to be optional, `jti` should be mandatory. This would fit with normal usage of the `kid` header claim, and prevents a breaking change to the spec around `jti`. (Similarly, `groups` must remain mandatory, as making it optional risks tokens without groups being rejected by implementations following earlier specs, and because the mp-jwt token doesn't include a version claim, there's no way to know which rule set to apply to an incoming token).

**Comment by chunlongliang-ibm** [comment]
@ayoho can you fix the tokenReuse option in mpJwt? If tokenreuse is false, and if jti is missing or reused, mak eit an error

**Comment by chunlongliang-ibm** [comment]
@ayoho you probably need a new config to have the behavior described above, an error if jti is missing or jti reused. The current tokenReuse only checks if jti is reused, and does not check if jti exists. The current default behavior should not be changed

**Comment by NottyCode** [comment]
@chunlongliang-ibm tokenReuse is set to true in metatype: https://openliberty.io/docs/ref/config/#mpJwt.html

So I don't understand your comment saying that it is disabled. I also don't understand why tokenReuse would affect validation that the JWT contains the spec required claims.

**Comment by chunlongliang-ibm** [comment]
To be compliant with spec, we will reject JWT that does not include 'jti', but we will provide an option not to check 'jti' existence.

**Comment by chunlongliang-ibm** [comment]
please assign me as code reviewer when code is ready.

**Comment by yeekangc** [comment]
This maps to https://bigblue.aha.io/features/SEC-25

**Comment by tevans78** [comment]
Still under consideration

**Comment by teddyjtorres** [comment]
The kid header is currently a recommended header, not a required one per https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/interoperability.asciidoc#recommended-mp-jwt-headers , where it states,

> Recommended MP-JWT header
>
>kid
>This JOSE header parameter is a hint indicating which key was used to secure the JWT. RFC7515, Section-4.1.4. It may need to be set if the verification key is in the JWK format.

The jti claim is currently a recommended claim, not a required one per https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/interoperability.asciidoc#recommended-mp-jwt-claims , where it states,

> Recommended MP-JWT claims
>
>jti
>Provides a unique identifier for the JWT. The identifier value MUST be assigned in a manner that ensures that there is a negligible probability that the same value will be accidentally assigned to a different data object; if the application uses multiple issuers, collisions MUST be prevented among values produced by different issuers as well. The "jti" claim can be used to prevent the JWT from being replayed. The "jti" value is a case-sensitive string. RFC7519, Section 4.1.7

**Comment by BarDweller** [comment]
Wow.. I thought this issue had retired and gone to live somewhere quietly in a nice log cabin in a desolate mountain range surrounded by trees and nature. 

The important part here is that the MP-JWT spec has been altered in the 2 and a bit years since I raised this Issue. 

At the time of raising, those headers were required. It looks like the change to the mp-jwt spec was made via https://github.com/eclipse/microprofile-jwt-auth/pull/191 which has additional comments re kid/jti that may need to be considered within the context of Liberty's usage, before deciding to omit them or not (specifically that kid, although optional, may still be required if the verify type is jwk. I forget what the verify type can be configured as in Liberty, as it's been such a long while since I raised this that I no longer work for IBM ;) )
